server/src/main/java/org/elasticsearch/search/MultiValueMode.java

```diff
@@ -63,16 +58,38 @@ protected long pick(SortedNumericDocValues values) throws IOException {
         }
 
         @Override
-        protected long pick(SortedNumericDocValues values, long missingValue, DocIdSetIterator docItr, int startDoc, int endDoc) throws IOException {
+        protected long pick(SortedNumericDocValues values, long missingValue, DocIdSetIterator docItr, int startDoc, int endDoc, Integer maxChildren) throws IOException {
             int totalCount = 0;
             long totalValue = 0;
-            for (int doc = startDoc; doc < endDoc; doc = docItr.nextDoc()) {
-                if (values.advanceExact(doc)) {
-                    final int count = values.docValueCount();
-                    for (int index = 0; index < count; ++index) {
-                        totalValue += values.nextValue();
+            if (maxChildren == null) {
```

Comment:
I think you can do:
````
int count = 0;
for (int doc = startDoc; doc < endDoc; doc = docItr.nextDoc()) {
  if (count++ > maxChildren) {
    break;
  }
....
````
We match the children in the same order than what was indexed so you can just count the number of matching children for the current parent document and stops when you have more than `maxChildren`.

Reply:
This `docItr` working `reverse`.
First item is actually  last item on index time.
And we cannot estimate how many item are in it.

It could not work.

I mentioned it in my previous comment.

Reply:
Right, sorry I forgot that we reverse to put all parents after the children. Considering that this new option is useful only if you order the nested documents client side I wonder if we could just document that the selection for `maxChildren` starts from the end ? It's not ideal in terms of API but we would not need to do the filtering in two passes so it would be efficient.

Reply:
Yes, Its more efficient but really hard to understand even with documentation.